### PR TITLE
PIE-2499: Create test ownership page

### PR DESCRIPTION
**Pull request by KatieWright26**
screencapture-buildkite-localhost-docs-test-analytics-test-ownership-2024-06-05-15_27_43

### CODE REVIEW

File: pages/test_analytics/test_ownership.md
Comment (KatieWright26) [reply pages/test_analytics/test_ownership.md:25]: Lol oops \*insert github blank teams rules\*

File: pages/test_analytics/test_ownership.md
Comment (jameshill) [inline pages/test_analytics/test_ownership.md:4]: I reckon we should do a little marketing type explanation around the problem and the solution, something like... 

```suggestion

Test ownership is critical to adopting a healthy testing culture at your organization. By defining test owners, your teams will become accountable for maintaining a fast and reliable test suite, ensuring confidence when you deploy your code.

```

This was a first pass crack at it... @mbelton-buildkite might have good words, but he's also likely slammed so we can always iterate.

File: pages/test_analytics/test_ownership.md
Comment (jameshill) [inline pages/test_analytics/test_ownership.md:5]: What do you think about dropping the _per-suite_ instructions here, and we can talk about that requirement later in the docs... mainly because there's nothing stopping someone using the same file uploaded to multiple suites. 

Something like this perhaps:

```suggestion
Test ownership is managed via team assignments in a TESTOWNER file. The team that is the default owner of a test [will be automatically assigned flaky tests](/docs/test-analytics/flaky-test-assignment) to triage.
```

File: pages/test_analytics/test_ownership.md
Comment (jameshill) [inline pages/test_analytics/test_ownership.md:19]: 💚

File: pages/test_analytics/test_ownership.md
Comment (jameshill) [inline pages/test_analytics/test_ownership.md:23]: it would be cool if we could link to the one-exception and put it at the bottom of the doc under some kinda troubleshooting section, that way we get to keep them on the happy path.

File: pages/test_analytics/test_ownership.md
Comment (jameshill) [inline pages/test_analytics/test_ownership.md:96]: i always get confused with case types... is snake-case `snake_case`... for example is it `test-analytics` or `test_analytics` and which one do we match? i think it's just the slug yeah?

File: pages/test_analytics/test_ownership.md
Comment (jameshill) [inline pages/test_analytics/test_ownership.md:102]: I had to do some mental gymnastics on this sentence, talking about permissions is damn hard. 

How about?

> Teams listed in the the TESTOWNERS must have permission to access the suite before ownership records are created.

🤷🏻 i hate talking about permissions 😭 it breaks my brain.

File: pages/test_analytics/test_ownership.md
Comment (jameshill) [inline pages/test_analytics/test_ownership.md:109]: We could say something about multiple suites using the same TESTOWNERS file. 

```suggestion
### FAQ

#### Can multiple suites use the one TESTOWNERS
Yes, there's nothing stopping you from uploading the same file to multiple suites. 
```

**Review by jameshill — COMMENTED** [review]
Great work @KatieWright26! 

Nothing is blocking on this, just one questions about snake case etc and I've added some suggestions to help sell the feature a little and make it a little easier to parse.

File: pages/test_analytics/test_ownership.md
Comment (jasmine-q) [reply pages/test_analytics/test_ownership.md:96]: oh it should be kebab case! Snake case is all lower case with underscores, kebab case uses lowercase with dashes

File: pages/test_analytics/test_ownership.md
Comment (KatieWright26) [reply pages/test_analytics/test_ownership.md:5]: Thanks I'll use that!

File: pages/test_analytics/test_ownership.md
Comment (KatieWright26) [reply pages/test_analytics/test_ownership.md:96]: AH it's kebab-case, not snake-case

File: pages/test_analytics/test_ownership.md
Comment (KatieWright26) [reply pages/test_analytics/test_ownership.md:23]: Yeah I think this section could just be moved down now as we have the example TESTOWNER file that describes the rules specifically now anyway.

**Review by jasmine-q — APPROVED** [review]

File: pages/test_analytics/test_ownership.md
Comment (jameshill) [reply pages/test_analytics/test_ownership.md]: i didn't even know it was called kebab case 🤣 i just thought it was dasherized.

**Review by jameshill — APPROVED** [review]
yeah this is looking great, I'm going to export it locally and share it with some beta testing customers to see how they go with the instructions.

**Review by gilesgas — CHANGES_REQUESTED** [review]
Reviewing now...

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:4]: Test Analytics has its own 'Access controls' docs page for this, with relevant links through to the general platform-level docs about teams, currently in the Pipelines section of the Buildkite Docs, but which should eventually be pulled out into its own 'Platform' docs section.

It'd be more sensible for this link to lead to the Test Analyics page equivalent first, since it also talks about test suite-level permissions.

```suggestion
Customers on the [Pro and Enterprise plans](https://buildkite.com/pricing) can assign test ownership to [Teams](/docs/test-analytics/permissions).
```

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:80]: Add gaps under headings to make the source content a little more readable.

```suggestion
### Can multiple suites use the same TESTOWNER file?

Yes, there's nothing stopping you from uploading the same file to multiple suites. A suite only has one active TESTOWNER file at a time.
```

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:54]: Is there a reason for the double-space here?

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:83]:
```suggestion
### Can more than one team own a test?

More than 1 team may own a test, and the order of teams in your TESTOWNER file is important. The first team listed will be the default owner, and they will be auto-assigned to the test if it flakes. Any team with suite access can override this auto-assignment.
```

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:86]:
```suggestion
### Why aren't my teams being assigned ownership over my tests?

A TESTOWNER file uses Buildkite team slugs instead of user names. Your team slug will be your team name in kebab-case. You can view your teams in your organization settings, or fetch them from our API:
```

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:2]:
```suggestion
# Test ownership

Test ownership is critical to adopting a healthy testing culture at your organization. By defining test owners, your teams will become accountable for maintaining a fast and reliable test suite, ensuring confidence when you deploy your code.
```

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [reply pages/test_analytics/test_ownership.md:5]: Looks like this was moved to the top. All good.

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:33]: Minor tweaks...

```suggestion
# These teams will be the test owners for everything in this
# repository. While both teams own the test, the first team listed will
# become the default owner for tests in this directory. Unless
# a later match takes precedence, team-slug-1 will have any
# flaky tests assigned to them by default.
* team-slug-1 team-slug-2
```

Also, couple of queries:

1. What was the reason for the extra space between the `*` and `team-slug-1`? If this functionally does nothing, I'd suggest removing it (as I've suggested above). Otherwise, please feel free to add it back in, with an explanation as to why the extra space is important.
2. When you say "a later match takes precedence", what exactly is being matched here and by what? (Sorry for my ignorance.)

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:38]: Minor tweaks ... please correct if my interpretation is incorrect ...

```suggestion
# Any test file ending with _spec.rb will be assigned
# to the test-analytics team and not team-slug-1.
# The last matching team takes precedence.
*_spec.rb test-analytics #This is an inline comment.
```

A couple of comments/questions:

1. Removed extra space (as per my previous comment).
1. Can multiple teams be specified against this file and if so, what's the behaviour in doing so?

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:40]: This line should probably be moved to after `# This is a comment.` above.
Then the 2nd sentence can be deleted as it's already covered by the following line.

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:49]:
```suggestion
# In this example, the packages team owns any files in the
# spec/packages/ directory at the root of the test directory and
# any of its subdirectories.
/spec/packages/ packages
```

**Review by gilesgas — APPROVED** [review]
All this looks pretty good @KatieWright26 !
Thanks very much for preparing this. I just had a few minor comments and suggested changes to make throughout it.

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:89]: Just a few more things I forgot about...

I don't think you need terminating `</li>` elements without a starting one. Also, this Markdown list syntax doesn't require any additional HTML syntax when it's rendered into HTML.

```suggestion
- [List teams from REST API](/docs/apis/rest_api/teams)
- [List teams from GraphQL API](/docs/apis/graphql/schemas/object/team)
```

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:98]:
```suggestion
Additionally, the teams listed in your TESTOWNERS file must have [permission to access the suite](/docs/test-analytics/permissions#manage-teams-and-permissions) before ownership records are created.
```

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [inline pages/test_analytics/test_ownership.md:108]:
```suggestion
# In a regular .gitignore or COEDOWNER file, the following block would
# set the test-analytics team as the owner of any file in the `/specs`
```

File: pages/test_analytics/test_ownership.md
Comment (KatieWright26) [reply pages/test_analytics/test_ownership.md:38]: The suggested change isn't quite right; we might have touched on this in slack already, but to clarify:

> Can multiple teams be specified against this file and if so, what's the behaviour in doing so?

Yes, multiple teams can own the same test location. The team listed first will be the "default owner" of the test, meaning when a test flakes they alone will be auto-assigned to fix it. There's more detail in the section above starting with `# These teams will be the owners for everything in`

> Is this configuration meant to override the * team-name-style configurations above?

Correct, the order in this entire file is important. If our file was just:
```bash
*              team-slug-1 team-slug-2
*_spec.rb      test-analytics
```
We assign ownership of all tests matching *_spec.rb` to the `test-analytics` team. Then, excluding tests matching `*_spec.rb`, we assign all tests (`*`) to `team-slug-1` and `team-slug-2`.

File: pages/test_analytics/test_ownership.md
Comment (KatieWright26) [reply pages/test_analytics/test_ownership.md:33]: The weird spacing has come from me copying over [Github's instructions for creating a CODEOWNER file](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners#example-of-a-codeowners-file) and overwriting it to match what we have. TESTOWNERSHIP uses the same rules as those set out for githubs `.gitignore` and `CODEOWNER` files, and I wanted to reuse something familiar. My assumption re the spacing is to try and show the difference between the location and the owner slugs. It's also how this [kind of file is usually laid out in real life](https://github.com/buildkite/buildkite/blob/main/.github/CODEOWNERS), but agree spacing is inconsistent e.g.
```bash
/docs docs docs-1
# as opposed to
/docs          docs docs-1
```

> When you say "a later match takes precedence", what exactly is being matched here and by what?

I think I answered this in https://github.com/buildkite/docs/pull/2813#discussion_r1641989485 by let me know if it's still unclear!

File: pages/test_analytics/test_ownership.md
Comment (KatieWright26) [reply pages/test_analytics/test_ownership.md:54]: Answered in https://github.com/buildkite/docs/pull/2813#discussion_r1641993258

**Comment by KatieWright26** [comment]
👋 @gilesgas  I've addressed your comments are you ok for me to merge this?

**Review by gilesgas — CHANGES_REQUESTED** [review]
Making some rewording tweaks to the `TESTOWNERS` file after a discussion with @KatieWright26 before merging in.

**Review by gilesgas — APPROVED** [review]

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [reply pages/test_analytics/test_ownership.md]: Thanks! Fixed and reworded following a more detailed discussion with you :-)

File: pages/test_analytics/test_ownership.md
Comment (gilesgas) [reply pages/test_analytics/test_ownership.md]: Thanks! Fixed and reworded following a more detailed discussion with you :-)

**Pull request merged**
